// @rumess/nepse-api/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractPageRange = exports.extractAllPages = exports.retryWithBackoff = exports.retry = exports.TokenParser = exports.TokenManager = exports.RetryManager = exports.PaginationManager = exports.Nepse = exports.DummyIDManager = exports.IndexIDEnum = exports.apiEndpoints = void 0;
const Nepse_1 = require("./core/Nepse");
Object.defineProperty(exports, "Nepse", { enumerable: true, get: function () { return Nepse_1.Nepse; } });
const DummyIDManager_1 = require("./utils/DummyIDManager");
Object.defineProperty(exports, "DummyIDManager", { enumerable: true, get: function () { return DummyIDManager_1.DummyIDManager; } });
const PaginationManager_1 = require("./utils/PaginationManager");
Object.defineProperty(exports, "PaginationManager", { enumerable: true, get: function () { return PaginationManager_1.PaginationManager; } });
Object.defineProperty(exports, "extractAllPages", { enumerable: true, get: function () { return PaginationManager_1.extractAllPages; } });
Object.defineProperty(exports, "extractPageRange", { enumerable: true, get: function () { return PaginationManager_1.extractPageRange; } });
const RetryManager_1 = require("./utils/RetryManager");
Object.defineProperty(exports, "RetryManager", { enumerable: true, get: function () { return RetryManager_1.RetryManager; } });
Object.defineProperty(exports, "retry", { enumerable: true, get: function () { return RetryManager_1.retry; } });
Object.defineProperty(exports, "retryWithBackoff", { enumerable: true, get: function () { return RetryManager_1.retryWithBackoff; } });
const TokenManager_1 = require("./utils/TokenManager");
Object.defineProperty(exports, "TokenManager", { enumerable: true, get: function () { return TokenManager_1.TokenManager; } });
const TokenParser_1 = require("./utils/TokenParser");
Object.defineProperty(exports, "TokenParser", { enumerable: true, get: function () { return TokenParser_1.TokenParser; } });
// Export API endpoints
var apiEndpoints_json_1 = require("./data/apiEndpoints.json");
Object.defineProperty(exports, "apiEndpoints", { enumerable: true, get: function () { return __importDefault(apiEndpoints_json_1).default; } });
// Export enums
var enums_1 = require("./enums");
Object.defineProperty(exports, "IndexIDEnum", { enumerable: true, get: function () { return enums_1.IndexIDEnum; } });
// Export errors
__exportStar(require("./errors"), exports);
// Default export
exports.default = Nepse_1.Nepse;
