// @rumess/nepse-api/dist/cli/getFloorsheet.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFloorSheet = getFloorSheet;
const cli_progress_1 = __importDefault(require("cli-progress"));
const Nepse_1 = require("../core/Nepse");
const PaginationManager_1 = require("../utils/PaginationManager");
const RetryManager_1 = require("../utils/RetryManager");
async function getFloorSheet(hideProgress = false) {
    const nepse = new Nepse_1.Nepse();
    nepse.setTLSVerification(false);
    let progressBar = null;
    let totalPages = 0;
    return PaginationManager_1.PaginationManager.extractAllPages({
        delayBetweenRequests: 100, // 100ms delay between requests
        onPageFetched: (page, data, totalItems) => {
            if (!hideProgress && progressBar) {
                progressBar.update(page + 1, { items: totalItems });
            }
        },
        onComplete: (totalItems) => {
            if (!hideProgress && progressBar) {
                progressBar.stop();
                console.log(`✅ Completed! Total items fetched: ${totalItems.length}`);
            }
        },
        fetchPage: async (page) => {
            // Initialize progress bar on first page
            if (page === 0 && !hideProgress) {
                const firstResponse = await RetryManager_1.RetryManager.executeWithFallback(async () => {
                    const response = await nepse.getFloorSheet({ page: 0 });
                    return {
                        totalPages: response.floorsheets.totalPages,
                        content: response.floorsheets.content,
                    };
                }, {
                    maxRetries: 3,
                    baseDelay: 5000,
                    backoffMultiplier: 1,
                    onRetry: (attempt, error, delay) => {
                        console.error(`\nError fetching page 1:`, error);
                        console.log(`\nRetrying page 1 in ${delay / 1000} seconds... (Attempt ${attempt}/3)`);
                    },
                    onMaxRetriesExceeded: (_error, attempts) => {
                        console.error(`\nFailed to fetch page 1 after ${attempts} attempts. Skipping...`);
                    },
                });
                if (firstResponse === null) {
                    return {
                        totalPages: 0,
                        content: [],
                    };
                }
                totalPages = firstResponse.totalPages;
                // Initialize progress bar
                progressBar = new cli_progress_1.default.SingleBar({
                    format: 'Downloading floorsheet |{bar}| {percentage}% | {value}/{total} pages | {items} items',
                    barCompleteChar: '\u2588',
                    barIncompleteChar: '\u2591',
                    hideCursor: true,
                });
                progressBar.start(totalPages, 1, {
                    items: firstResponse.content.length,
                });
                return firstResponse;
            }
            const result = await RetryManager_1.RetryManager.executeWithFallback(async () => {
                const response = await nepse.getFloorSheet({ page });
                return {
                    totalPages: response.floorsheets.totalPages,
                    content: response.floorsheets.content,
                };
            }, {
                maxRetries: 3,
                baseDelay: 5000, // 5 seconds
                backoffMultiplier: 1, // No exponential backoff for API calls
                onRetry: (attempt, error, delay) => {
                    console.error(`\nError fetching page ${page + 1}:`, error);
                    console.log(`\nRetrying page ${page + 1} in ${delay / 1000} seconds... (Attempt ${attempt}/3)`);
                },
                onMaxRetriesExceeded: (_error, attempts) => {
                    console.error(`\nFailed to fetch page ${page + 1} after ${attempts} attempts. Skipping...`);
                },
            });
            if (result === null) {
                // Return empty response if all retries failed
                return {
                    totalPages: 0,
                    content: [],
                };
            }
            return result;
        },
    });
}
