// @rumess/nepse-api/dist/core/BaseNepse.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseNepse = void 0;
const axios_1 = __importDefault(require("axios"));
const apiEndpoints_json_1 = __importDefault(require("../data/apiEndpoints.json"));
const headers_json_1 = __importDefault(require("../data/headers.json"));
const errors_1 = require("../errors");
/**
 * Base class for Nepse API
 */
class BaseNepse {
    constructor() {
        this.tlsVerify = false;
        this.companySymbolIdKeymap = null;
        this.securitySymbolIdKeymap = null;
        this.companyList = null;
        this.securityList = null;
        this.sectorScrips = null;
        this.floorSheetSize = 500;
        this.baseUrl = 'https://nepalstock.com.np'; // Try alternative domain
        this.apiEndpoints = apiEndpoints_json_1.default;
        this.headers = headers_json_1.default;
        this.headers.Host = this.baseUrl.replace('https://', '');
        this.headers.Referer = this.baseUrl.replace('https://', '');
        this.initClient();
    }
    /**
     * Set the TLS verification
     * @param verify - Whether to verify the TLS certificate
     */
    setTLSVerification(verify) {
        this.tlsVerify = verify;
        this.initClient();
    }
    /**
     * Initialize the client
     */
    initClient() {
        this.client = axios_1.default.create({
            baseURL: this.baseUrl,
            timeout: 100000, // Increased to match Python timeout
            httpsAgent: this.tlsVerify
                ? undefined
                : new (require('https').Agent)({
                    requestCert: false,
                    rejectUnauthorized: false,
                }),
            headers: this.headers,
        });
    }
    /**
     * Get the full URL
     * @param apiUrl - The API URL
     * @returns The full URL
     */
    getFullUrl(apiUrl) {
        return `${this.baseUrl}${apiUrl}`;
    }
    /**
     * Get the authorization headers
     * @param accessToken - The access token
     * @returns The authorization headers
     */
    getAuthorizationHeaders(accessToken) {
        return {
            Authorization: `Salter ${accessToken}`,
            'Content-Type': 'application/json',
            ...this.headers,
        };
    }
    /**
     * Handle the response
     * @param response - The response
     * @returns The response data
     */
    handleResponse(response) {
        if (response.status >= 200 && response.status < 300) {
            return response.data;
        }
        else if (response.status >= 400 && response.status < 500) {
            throw new errors_1.NepseInvalidClientRequest(`Client error: ${response.status}`);
        }
        else if (response.status >= 500) {
            throw new errors_1.NepseInvalidServerResponse(`Server error: ${response.status}`);
        }
        else {
            throw new errors_1.NepseNetworkError(`Unexpected response: ${response.status}`);
        }
    }
    /**
     * Request a GET API
     * @param url - The URL
     * @param includeAuth - Whether to include the authorization headers
     * @returns The response data
     */
    async requestGETAPI(url, includeAuth = true) {
        try {
            const config = {};
            if (includeAuth) {
                const accessToken = await this.tokenManager.getAccessToken();
                config.headers = this.getAuthorizationHeaders(accessToken);
            }
            else {
                config.headers = this.headers;
            }
            const response = await this.client.get(url, config);
            return this.handleResponse(response);
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                if (error.response) {
                    throw new errors_1.NepseNetworkError(`HTTP ${error.response.status}: ${error.response.statusText}`);
                }
                else if (error.request) {
                    throw new errors_1.NepseNetworkError(`Network error: No response received: ${error.message}`);
                }
                else {
                    throw new errors_1.NepseNetworkError(`Request error: ${error.message}`);
                }
            }
            else {
                if (error instanceof errors_1.NepseNetworkError) {
                    throw error;
                }
                throw new errors_1.NepseNetworkError(`Unknown error: ${String(error)}`);
            }
        }
    }
    /**
     * Request a POST API
     * @param url - The URL
     * @param payload - The payload
     * @param includeAuth - Whether to include the authorization headers
     * @returns The response data
     */
    async requestPOSTAPI(url, payload, includeAuth = true) {
        try {
            const config = {};
            if (includeAuth) {
                const accessToken = await this.tokenManager.getAccessToken();
                config.headers = this.getAuthorizationHeaders(accessToken);
            }
            else {
                config.headers = this.headers;
            }
            const response = await this.client.post(url, payload, config);
            return this.handleResponse(response);
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                if (error.response) {
                    throw new errors_1.NepseNetworkError(`HTTP ${error.response.status}: ${error.response.statusText}`);
                }
                else if (error.request) {
                    throw new errors_1.NepseNetworkError('Network error: No response received');
                }
                else {
                    throw new errors_1.NepseNetworkError(`Request error: ${error.message}`);
                }
            }
            else {
                throw new errors_1.NepseNetworkError(`Unknown error: ${String(error)}`);
            }
        }
    }
    /**
     * Get the company ID keymap
     * @returns The company ID keymap
     */
    async getCompanyIdKeymap() {
        if (!this.companySymbolIdKeymap) {
            this.companySymbolIdKeymap = new Map((await this.getCompanyList()).map((company) => [
                company.symbol,
                company.id,
            ]));
        }
        return this.companySymbolIdKeymap;
    }
    /**
     * Get the security symbol ID keymap
     * @returns The security symbol ID keymap
     */
    async getSecuritySymbolIdKeymap(force) {
        if (!this.securitySymbolIdKeymap || force) {
            this.securitySymbolIdKeymap = new Map((await this.getSecurityList(force)).map((security) => [
                security.symbol,
                security.id,
            ]));
        }
        return this.securitySymbolIdKeymap;
    }
    /**
     * Get the security ID keymap
     * @returns The security ID keymap
     */
    async getSecurityIdKeymap(force) {
        if (!this.securitySymbolIdKeymap || force) {
            this.securitySymbolIdKeymap = new Map((await this.getSecurityList(force)).map((security) => [
                security.symbol,
                security.id,
            ]));
        }
        return this.securitySymbolIdKeymap;
    }
}
exports.BaseNepse = BaseNepse;
