// @rumess/nepse-api/dist/errors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NepseTokenExpired = exports.NepseInvalidServerResponse = exports.NepseInvalidClientRequest = exports.NepseNetworkError = exports.NepseError = void 0;
class NepseError extends Error {
    constructor(message) {
        super(message);
        this.name = 'NepseError';
    }
}
exports.NepseError = NepseError;
class NepseNetworkError extends NepseError {
    constructor(message) {
        super(message);
        this.name = 'NepseNetworkError';
    }
}
exports.NepseNetworkError = NepseNetworkError;
class NepseInvalidClientRequest extends NepseError {
    constructor(message) {
        super(message);
        this.name = 'NepseInvalidClientRequest';
    }
}
exports.NepseInvalidClientRequest = NepseInvalidClientRequest;
class NepseInvalidServerResponse extends NepseError {
    constructor(message) {
        super(message);
        this.name = 'NepseInvalidServerResponse';
    }
}
exports.NepseInvalidServerResponse = NepseInvalidServerResponse;
class NepseTokenExpired extends NepseError {
    constructor(message) {
        super(message);
        this.name = 'NepseTokenExpired';
    }
}
exports.NepseTokenExpired = NepseTokenExpired;
