// @rumess/nepse-api/dist/utils/DummyIDManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DummyIDManager = exports.BaseDummyIDManager = void 0;
class BaseDummyIDManager {
    constructor(marketStatusFunction) {
        this.data = null;
        this.dummyId = null;
        this.dateStamp = null;
        this.marketStatusFunction = null;
        this.setMarketStatusFunction(marketStatusFunction || null);
    }
    setMarketStatusFunction(func) {
        this.marketStatusFunction = func;
        this.data = null;
    }
    convertToDateTime(dateTimeStr) {
        return new Date(dateTimeStr);
    }
    toString() {
        return `<Dummy ID: ${this.dummyId}, Date: ${this.dateStamp}>`;
    }
}
exports.BaseDummyIDManager = BaseDummyIDManager;
class DummyIDManager extends BaseDummyIDManager {
    constructor() {
        super(...arguments);
        this.updateStarted = false;
        this.updatePromise = null;
    }
    async populateData(force = false) {
        const today = new Date();
        if (this.data === null || force) {
            if (!this.updateStarted) {
                this.updateStarted = true;
                this.updatePromise = this.performUpdate();
                await this.updatePromise;
                this.updateStarted = false;
                this.updatePromise = null;
                return;
            }
            await this.updatePromise;
        }
        if (this.dateStamp && this.dateStamp.getDate() < today.getDate()) {
            if (this.updateStarted) {
                await this.updatePromise;
            }
            else {
                this.updateStarted = true;
                this.updatePromise = this.performUpdate();
                await this.updatePromise;
                this.updateStarted = false;
                this.updatePromise = null;
            }
        }
    }
    async performUpdate() {
        if (!this.marketStatusFunction) {
            throw new Error('Market status function not set');
        }
        const newData = await this.marketStatusFunction();
        const today = new Date();
        const newConvertedDate = this.convertToDateTime(newData.asOf);
        if (newConvertedDate.getDate() === today.getDate()) {
            this.data = newData;
            this.dummyId = this.data.id;
            this.dateStamp = newConvertedDate;
        }
        else {
            this.data = newData;
            this.dummyId = this.data.id;
            this.dateStamp = today;
        }
    }
    async getDummyID() {
        await this.populateData();
        if (this.dummyId === null) {
            throw new Error('Dummy ID not available');
        }
        return this.dummyId;
    }
}
exports.DummyIDManager = DummyIDManager;
