// @rumess/nepse-api/dist/utils/RetryManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RetryManager = void 0;
exports.retry = retry;
exports.retryWithBackoff = retryWithBackoff;
class RetryManager {
    /**
     * Executes a function with retry logic
     * @param fn The function to execute
     * @param options Retry configuration options
     * @returns Promise that resolves with the function result or rejects after max retries
     */
    static async execute(fn, options = {}) {
        const config = { ...RetryManager.DEFAULT_OPTIONS, ...options };
        let lastError;
        for (let attempt = 0; attempt <= config.maxRetries; attempt++) {
            try {
                return await fn();
            }
            catch (error) {
                lastError = error;
                // Check if we should retry this error
                if (!config.retryCondition(error)) {
                    throw error;
                }
                // If this was the last attempt, don't retry
                if (attempt === config.maxRetries) {
                    config.onMaxRetriesExceeded(error, attempt + 1);
                    throw error;
                }
                // Calculate delay with exponential backoff
                const delay = Math.min(config.baseDelay * Math.pow(config.backoffMultiplier, attempt), config.maxDelay);
                config.onRetry(attempt + 1, error, delay);
                await RetryManager.sleep(delay);
            }
        }
        throw lastError;
    }
    /**
     * Executes a function with retry logic and continues on failure
     * @param fn The function to execute
     * @param options Retry configuration options
     * @returns Promise that resolves with the function result or null if all retries failed
     */
    static async executeWithFallback(fn, options = {}) {
        try {
            return await RetryManager.execute(fn, options);
        }
        catch {
            return null;
        }
    }
    /**
     * Utility method to sleep for a given duration
     * @param ms Duration in milliseconds
     */
    static sleep(ms) {
        return new Promise((resolve) => setTimeout(resolve, ms));
    }
}
exports.RetryManager = RetryManager;
RetryManager.DEFAULT_OPTIONS = {
    maxRetries: 3,
    baseDelay: 1000, // 1 second
    maxDelay: 30000, // 30 seconds
    backoffMultiplier: 2,
    retryCondition: () => true, // Retry on any error by default
    onRetry: () => { },
    onMaxRetriesExceeded: () => { },
};
/**
 * Convenience function for simple retry operations
 * @param fn The function to execute
 * @param maxRetries Maximum number of retry attempts
 * @param delay Delay between retries in milliseconds
 * @returns Promise that resolves with the function result
 */
async function retry(fn, maxRetries = 3, delay = 1000) {
    return RetryManager.execute(fn, {
        maxRetries,
        baseDelay: delay,
        backoffMultiplier: 1, // No exponential backoff for simple retry
    });
}
/**
 * Convenience function for retry with exponential backoff
 * @param fn The function to execute
 * @param maxRetries Maximum number of retry attempts
 * @param baseDelay Base delay in milliseconds
 * @returns Promise that resolves with the function result
 */
async function retryWithBackoff(fn, maxRetries = 3, baseDelay = 1000) {
    return RetryManager.execute(fn, {
        maxRetries,
        baseDelay,
        backoffMultiplier: 2,
    });
}
