// @rumess/nepse-api/dist/utils/TokenParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenParser = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
class TokenParser {
    constructor() {
        this.wasmModule = null;
        this.wasmMemory = null;
        this.initializationPromise = null;
        this.initializationPromise = this.initializeWasm();
    }
    async initializeWasm() {
        try {
            const wasmPath = (0, path_1.join)(__dirname, '../data/css.wasm');
            const wasmBuffer = (0, fs_1.readFileSync)(wasmPath);
            const wasmModule = await WebAssembly.instantiate(wasmBuffer);
            this.wasmModule = wasmModule.instance;
            this.wasmMemory = wasmModule.instance.exports
                .memory;
        }
        catch (error) {
            console.error('Failed to initialize WebAssembly module:', error);
            throw new Error('WebAssembly initialization failed');
        }
    }
    invokeWasmFunction(functionName, args) {
        if (!this.wasmModule) {
            throw new Error('WebAssembly module not initialized');
        }
        const func = this.wasmModule.exports[functionName];
        if (!func) {
            throw new Error(`WebAssembly function ${functionName} not found`);
        }
        return func(...args);
    }
    async parseTokenResponse(tokenResponse) {
        if (this.initializationPromise) {
            await this.initializationPromise;
        }
        if (!this.wasmModule) {
            throw new Error('WebAssembly module not initialized');
        }
        // First set of calculations
        const n = this.invokeWasmFunction('cdx', [
            tokenResponse.salt1,
            tokenResponse.salt2,
            tokenResponse.salt3,
            tokenResponse.salt4,
            tokenResponse.salt5,
        ]);
        const l = this.invokeWasmFunction('rdx', [
            tokenResponse.salt1,
            tokenResponse.salt2,
            tokenResponse.salt4,
            tokenResponse.salt3,
            tokenResponse.salt5,
        ]);
        const o = this.invokeWasmFunction('bdx', [
            tokenResponse.salt1,
            tokenResponse.salt2,
            tokenResponse.salt4,
            tokenResponse.salt3,
            tokenResponse.salt5,
        ]);
        const p = this.invokeWasmFunction('ndx', [
            tokenResponse.salt1,
            tokenResponse.salt2,
            tokenResponse.salt4,
            tokenResponse.salt3,
            tokenResponse.salt5,
        ]);
        const q = this.invokeWasmFunction('mdx', [
            tokenResponse.salt1,
            tokenResponse.salt2,
            tokenResponse.salt4,
            tokenResponse.salt3,
            tokenResponse.salt5,
        ]);
        // Second set of calculations
        const a = this.invokeWasmFunction('cdx', [
            tokenResponse.salt2,
            tokenResponse.salt1,
            tokenResponse.salt3,
            tokenResponse.salt5,
            tokenResponse.salt4,
        ]);
        const b = this.invokeWasmFunction('rdx', [
            tokenResponse.salt2,
            tokenResponse.salt1,
            tokenResponse.salt3,
            tokenResponse.salt4,
            tokenResponse.salt5,
        ]);
        const c = this.invokeWasmFunction('bdx', [
            tokenResponse.salt2,
            tokenResponse.salt1,
            tokenResponse.salt4,
            tokenResponse.salt3,
            tokenResponse.salt5,
        ]);
        const d = this.invokeWasmFunction('ndx', [
            tokenResponse.salt2,
            tokenResponse.salt1,
            tokenResponse.salt4,
            tokenResponse.salt3,
            tokenResponse.salt5,
        ]);
        const e = this.invokeWasmFunction('mdx', [
            tokenResponse.salt2,
            tokenResponse.salt1,
            tokenResponse.salt4,
            tokenResponse.salt3,
            tokenResponse.salt5,
        ]);
        const accessToken = tokenResponse.accessToken;
        const refreshToken = tokenResponse.refreshToken;
        // Parse access token
        const parsedAccessToken = accessToken.substring(0, n) +
            accessToken.substring(n + 1, l) +
            accessToken.substring(l + 1, o) +
            accessToken.substring(o + 1, p) +
            accessToken.substring(p + 1, q) +
            accessToken.substring(q + 1);
        // Parse refresh token
        const parsedRefreshToken = refreshToken.substring(0, a) +
            refreshToken.substring(a + 1, b) +
            refreshToken.substring(b + 1, c) +
            refreshToken.substring(c + 1, d) +
            refreshToken.substring(d + 1, e) +
            refreshToken.substring(e + 1);
        return [parsedAccessToken, parsedRefreshToken];
    }
}
exports.TokenParser = TokenParser;
