// @rumess/nepse-api/dist/utils/TokenManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenManager = exports.BaseTokenManager = void 0;
const TokenParser_1 = require("./TokenParser");
class BaseTokenManager {
    constructor(nepse) {
        this.MAX_UPDATE_PERIOD = 45; // seconds
        this.accessToken = null;
        this.refreshToken = null;
        this.tokenTimeStamp = null;
        this.salts = null;
        this.tokenUrl = '/api/authenticate/prove';
        this.refreshUrl = '/api/authenticate/refresh-token';
        this.nepse = nepse;
        this.tokenParser = new TokenParser_1.TokenParser();
    }
    isTokenValid() {
        if (!this.tokenTimeStamp)
            return false;
        return (Math.floor(Date.now() / 1000) - this.tokenTimeStamp <
            this.MAX_UPDATE_PERIOD);
    }
    async getValidTokenFromJSON(tokenResponse) {
        const salts = [
            parseInt(tokenResponse.salt1.toString()),
            parseInt(tokenResponse.salt2.toString()),
            parseInt(tokenResponse.salt3.toString()),
            parseInt(tokenResponse.salt4.toString()),
            parseInt(tokenResponse.salt5.toString()),
        ];
        const [parsedAccessToken, parsedRefreshToken] = await this.tokenParser.parseTokenResponse(tokenResponse);
        return [
            parsedAccessToken,
            parsedRefreshToken,
            Math.floor(tokenResponse.serverTime / 1000),
            salts,
        ];
    }
    toString() {
        if (this.accessToken && this.tokenTimeStamp) {
            return `Access Token: ${this.accessToken}\nRefresh Token: ${this.refreshToken}\nSalts: ${this.salts}\nTimeStamp: ${new Date(this.tokenTimeStamp * 1000).toISOString()}`;
        }
        return 'Token Manager Not Initialized';
    }
}
exports.BaseTokenManager = BaseTokenManager;
class TokenManager extends BaseTokenManager {
    constructor() {
        super(...arguments);
        this.updateStarted = false;
        this.updatePromise = null;
    }
    async getAccessToken() {
        if (this.isTokenValid() && this.accessToken) {
            return this.accessToken;
        }
        else {
            await this.update();
            if (!this.accessToken) {
                throw new Error('Failed to obtain access token');
            }
            return this.accessToken;
        }
    }
    async getRefreshToken() {
        if (this.isTokenValid() && this.refreshToken) {
            return this.refreshToken;
        }
        else {
            await this.update();
            if (!this.refreshToken) {
                throw new Error('Failed to obtain refresh token');
            }
            return this.refreshToken;
        }
    }
    getSalts() {
        return this.salts;
    }
    async update() {
        await this.setToken();
    }
    async setToken() {
        if (!this.updateStarted) {
            this.updateStarted = true;
            this.updatePromise = this.performTokenUpdate();
            await this.updatePromise;
            this.updateStarted = false;
            this.updatePromise = null;
        }
        else {
            await this.updatePromise;
        }
    }
    async performTokenUpdate() {
        const response = await this.nepse.requestGETAPI(this.tokenUrl, false);
        const jsonResponse = response;
        [this.accessToken, this.refreshToken, this.tokenTimeStamp, this.salts] =
            await this.getValidTokenFromJSON(jsonResponse);
    }
}
exports.TokenManager = TokenManager;
